### PR TITLE
cmake: automatically detect engines dir

### PR DESCRIPTION
**Pull request by chipitsine**
@levitte , can you have a look ? it is based on your approach

I suppose this PR will resolve https://github.com/gost-engine/engine/issues/60

### CODE REVIEW

**Comment by chipitsine** [comment]
@beldmit I agree that install directory only make sense for "install" target, not "build"
I'll think how to deal with that, thank for the tip

File: CMakeLists.txt [hunk 1]
```diff
@@ -125,8 +125,29 @@ set(GOST_12_SUM_SOURCE_FILES
...
+EXECUTE_PROCESS(COMMAND ${CMAKE_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/CMakeTmp/conftest
+  OUTPUT_VARIABLE OUTPUT
+  RESULT_VARIABLE RESULT)
+
+IF(RESULT)
+  MESSAGE(FATAL_ERROR "engines directory is not detected.\n RESULT=${RESULT} OUTPUT=${OUTPUT}")
+ENDIF(RESULT)
+
+STRING(REGEX MATCH "\"([^\"]+)\"" OUTPUT1 ${OUTPUT})
+STRING(REPLACE "\"" "" OPENSSL_ENGINES_INSTALL_DIR ${OUTPUT1})
```
[the remark this answers is not in the document]
Comment (chipitsine) [reply CMakeLists.txt]: OUTPUT=ENGINESDIR: "/usr/lib64/engines-1.1"
OUTPUT1="/usr/lib64/engines-1.1"
OPENSSL_ENGINES_INSTALL_DIR=/usr/lib64/engines-1.1

**Comment by chipitsine** [comment]
also, I'd like to use proper macros instead of ```-lcrypto -lssl``` (but I could not find any)

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (chipitsine) [reply CMakeLists.txt]: I could not get rid of quotes in a single pass.
maybe, I'll be more lucky next time

**Comment by levitte** [comment]
There is a cmake module to find openssl. https://cmake.org/cmake/help/v3.6/module/FindOpenSSL.html

**Comment by chipitsine** [comment]
yes, I did have a look at that module. haven't found how to say ```-lcrypto -lssl``` with those variables
any clue?

**Comment by levitte** [comment]
Have a look at this for inspiration on use.

https://stackoverflow.com/a/44846535/6108256

(i.e., *you* shouldn't specify `-lcrypto -lssl`, it seems that cmake does that job for you)

**Comment by chipitsine** [comment]
I did have a look already. It is good for typical cmake usage, i.e. generating makefiles.
What am I doing is not typical, I build and run program during cmake run

**Comment by DJm00n** [comment]
Maybe you could try to use [try_run CMake macro](https://cmake.org/cmake/help/v3.7/command/try_run.html) and use `RUN_OUTPUT_VARIABLE ` instead.
Look at [CMake Modules dir](https://github.com/Kitware/CMake/tree/master/Modules) for examples.

**Comment by chipitsine** [comment]
I'll try it and write here the result

**Comment by candrews** [comment]
I've requested that CMake's FindOpenSSL get the ability to tell us the OpenSSL engine directory: https://gitlab.kitware.com/cmake/cmake/issues/18525

**Comment by chipitsine** [comment]
Thank you for bringing this up:)

I'll do something regarding this PR soon

**Pull request closed without merging**